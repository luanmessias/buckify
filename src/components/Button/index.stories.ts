import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { Button } from './';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Base/Button',
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    variant: { 
        control: 'select',
        options: ['primary', 'outline', 'icon', 'iconPrimary', 'iconOutline'],
    },
    as: {
      control: 'select',
      options: ['button', 'link'],
    },
  },
  // Use `fn` to spy on the onClick arg, which will appear in the actions panel once invoked: https://storybook.js.org/docs/essentials/actions#action-args
  args: { onClick: fn() },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const ButtonPrimary: Story = {
  args: {
    variant: 'primary',
    label: 'Primary Button',
    as: 'button',
  },
};

export const ButtonOutline: Story = {
  args: {
    variant: 'outline',
    label: 'Outline Button',
    as: 'button',
  },
};

export const ButtonWithIcon: Story = {
  args: {
    variant: 'outline',
    label: 'Outline Button With Icon',
    as: 'button',
    icon: 'activity',
  },
};

export const ButtonWithBrand: Story = {
  args: {
    variant: 'outline',
    label: 'Outline Button With Icon',
    as: 'button',
    brand: 'google-g-logo',
  },
};

export const OnlyIcon: Story = {
  args: {
    variant: "icon",
    label: "",
    as: "button",
    icon: "activity",
    iconColor: "#ffffff"
  }
};

export const IconPrimary: Story = {
  args: {
    variant: "iconPrimary",
    label: "",
    as: "button",
    icon: "activity",
    iconColor: "#ffffff"
  }
};

export const iconOutline: Story = {
  args: {
    variant: "iconOutline",
    label: "",
    as: "button",
    icon: "activity",
    iconColor: "gray"
  }
};

export const BrandOnly: Story = {
  args: {
    variant: "icon",
    label: "",
    as: "button",
    brand: "google-g-logo"
  }
};

export const BrandPrimary: Story = {
  args: {
    variant: "iconPrimary",
    label: "",
    as: "button",
    brand: "google-g-logo"
  }
};

export const BrandOutline: Story = {
  args: {
    variant: "iconOutline",
    label: "",
    as: "button",
    brand: "google-g-logo"
  }
};

export const ButtonAsLink: Story = {
  args: {
    variant: "outline",
    label: "This Button Is a Link",
    as: "link",
    brand: "",
    icon: "link-2",
    href: "https://www.google.com/",
    targetBlank: true
  }
};