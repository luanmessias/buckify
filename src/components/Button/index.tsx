import React, { ComponentProps } from 'react'
import { tv, VariantProps } from 'tailwind-variants'
import Link from 'next/link'
import Image from 'next/image'
import { Icon } from '../Icon'
import dynamicIconImports from 'lucide-react/dynamicIconImports'
import { motion } from 'framer-motion'

const button = tv({
  base: 'flex items-center gap-4 w-full py-4 px-6 rounded text-base transition duration-default ease-in-out',
  variants: {
    variant: {
      primary: 'bg-teal-500 text-white hover:bg-teal-600',
      outline:
        'border border-gray-300 bg-white text-gray-500 hover:bg-gray-100',
      icon: 'py-4 px-4 text-zero',
      iconPrimary:
        'bg-teal-500 text-white py-4 px-4 text-zero hover:bg-teal-600',
      iconOutline:
        'border border-gray-300 bg-white text-gray-500 py-4 px-4 text-zero hover:bg-gray-100'
    }
  }
})

export type ButtonProps = ComponentProps<'button'> &
  VariantProps<typeof button> & {
    variant: 'primary' | 'outline' | 'icon' | 'iconPrimary' | 'iconOutline'
    label?: string
    icon?: keyof typeof dynamicIconImports
    iconColor?: string
    brand?: string
    onClick: () => void
    as?: 'button' | 'link'
    href?: string
    targetBlank?: boolean
  }

export const Button = ({
  variant = 'primary',
  label,
  as = 'button',
  href,
  icon,
  iconColor,
  ...props
}: ButtonProps) => {
  const Component: React.ElementType = as === 'link' ? Link : 'button'
  const iconColorStyle = iconColor ? iconColor : 'currentColor'
  return (
    <motion.div whileHover={{ scale: 1.1 }} whileTap={{ scale: 0.95 }}>
      <Component
        className={button({ variant })}
        {...(as === 'link' ? { href } : { as })}
        {...(as === 'link'
          ? { target: props.targetBlank ? '_blank' : undefined }
          : {})}
        {...props}
      >
        {props.brand && (
          <Image
            aria-hidden
            src={`./brands/${props.brand}.svg`}
            alt="File icon"
            width={28}
            height={28}
            unoptimized
          />
        )}

        {icon && <Icon name={icon} size={24} color={iconColorStyle} />}

        {label}
      </Component>
    </motion.div>
  )
}
