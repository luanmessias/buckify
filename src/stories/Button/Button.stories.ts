import type { Meta, StoryObj } from '@storybook/vue3'

import BaseButton from './'
import YtIcon from './img/icon.vue'

const meta = {
  title: 'Navigation/Button',
  component: BaseButton,
  tags: ['autodocs'],
  argTypes: {
    label: {
      description: 'The label of the button'
    },
    disabled: {
      description: 'Whether the button is disabled'
    },
    type: {
      control: 'select',
      description: 'The type of the button',
      options: ['button', 'submit', 'reset']
    },
    theme: {
      control: 'select',
      description: 'The theme of the button',
      options: ['primary', 'secondary', 'disabled']
    },
    size: {
      control: 'select',
      description: 'The size of the button',
      options: ['small', 'medium', 'large', 'full']
    },
    onClick: {
      action: 'clicked',
      description: 'Click of the button'
    },
    to: {
      description: 'The route to link to'
    }
  },
  args: {
    onClick: () => console.log('clicked'),
    label: 'Button'
  }
} satisfies Meta<typeof BaseButton>

export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {}

export const Secondary: Story = {
  args: {
    theme: 'secondary'
  }
}

export const WithIcon: Story = {
  render: (args) => ({
    components: { YtIcon, BaseButton },
    setup() {
      return {
        args
      }
    },
    template: `
      <BaseButton v-bind="args">
        <YtIcon />
      </BaseButton>
    `
  }),
  args: {
    theme: 'secondary'
  }
}

export const Disabled: Story = {
  args: {
    disabled: true,
    theme: 'disabled'
  }
}
