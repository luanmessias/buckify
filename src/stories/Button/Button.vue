<template>
    <button 
      :type="type"
      :class="`${styles.base} ${selectedTheme} ${selectedSize}`" 
      :theme="theme"
      :size="size"
      @click="onClick" 
    >
      <slot/>
      {{ label }}
    </button>
</template>

<script lang="ts" setup>
  import { computed } from 'vue';

  const props = defineProps<{
    label: string;
    type: 'button' | 'submit' | 'reset';
    theme: 'primary' | 'secondary';
    size: 'small' | 'medium' | 'large';
    onClick: () => void;
  }>();

  const styles = {
    base: 'flex items-center justify-center rounded-md text-sm h-14 transition-colors duration-200 ease-in-out',
    theme: {
      primary: 'bg-primary-500 hover:bg-primary-700 text-white',
      secondary: 'bg-white hover:bg-gray-200 text-neutral-500 border border-neutral-100',
    },
    size: {
      small: 'w-20',
      medium: 'w-40',
      large: 'w-60',
      full: 'w-full',
    },
  };


  const selectedTheme = computed(() => {
    const theme = props.theme as keyof typeof styles.theme;
    return styles.theme[theme];
  });

  const selectedSize = computed(() => {
    const size = props.size as keyof typeof styles.size;
    return styles.size[size];
  });

</script>