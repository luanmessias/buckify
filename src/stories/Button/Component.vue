<template>
  <RouterLink
    v-if="to && !disabled"
    :to="to"
    :type="type"
    :class="`${styles.base} ${selectedTheme} ${selectedSize}`"
    :theme="theme"
    :size="size"
  >
    <slot />
    {{ label }}
  </RouterLink>
  <button
    v-else-if="!to && !disabled"
    :type="type"
    :class="`${styles.base} ${selectedTheme} ${selectedSize}`"
    :theme="theme"
    :size="size"
    @click="onClick"
  >
    <slot />
    {{ label }}
  </button>
  <div
    v-else
    :class="`${styles.base} ${styles.theme.disabled} ${selectedSize}`"
    :size="size"
  >
    <slot />
    {{ label }}
  </div>
</template>

<script lang="ts" setup>
  import { computed } from 'vue';
import { RouterLink } from 'vue-router';

  const props = defineProps({
    label: {
      type: String,
      required: true
    },
    type: {
      type: String as () => 'button' | 'submit' | 'reset',
      default: 'button'
    },
    theme: {
      type: String as () => 'primary' | 'secondary' | 'disabled',
      default: 'primary'
    },
    size: {
      type: String as () => 'small' | 'medium' | 'large',
      default: 'medium'
    },
    disabled: {
      type: Boolean,
      default: false
    },
    to: {
      type: String,
    },
    onClick: {
      type: Function as unknown as () => () => void
    }
  })

  const styles = {
    base: 'flex items-center justify-center gap-1 rounded-md text-sm h-14 transition-colors duration-200 ease-in-out',
    theme: {
      primary: 'bg-primary-500 hover:bg-primary-700 text-white',
      secondary: 'bg-white hover:bg-gray-200 text-neutral-500 border border-neutral-100',
      disabled: 'bg-neutral-100 text-neutral-300 cursor-not-allowed'
    },
    size: {
      small: 'w-20',
      medium: 'w-40',
      large: 'w-60',
      full: 'w-full'
    }
  }

  const selectedTheme = computed(() => {
    const theme = props.theme as keyof typeof styles.theme
    return styles.theme[theme]
  })

  const selectedSize = computed(() => {
    const size = props.size as keyof typeof styles.size
    return styles.size[size]
  })
</script>
