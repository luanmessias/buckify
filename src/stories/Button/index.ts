import Button from './Component.vue';

export default Button;
