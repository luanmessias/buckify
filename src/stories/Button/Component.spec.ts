import { describe, expect, it } from "vitest";

import { mount } from "@vue/test-utils";

import Button from "./Component.vue";

const props = {
  label: 'Click me!',
};

const wrapper = mount(Button, { props });


describe('ButtonComponent', () => {
  describe.only('render', () => {
    it('should be a router link', async () => {
      const to = '/test';
      await wrapper.setProps({ to });

      expect(wrapper.html()).toContain(`href="${to}"`);
    });
  });

  describe('props', () => {
    it('should render label', () => {
      expect(wrapper.text()).toBe(props.label);
    });

    it('should render the given type', async () => {
      const type = 'submit';
      await wrapper.setProps({ type });
      
      expect(wrapper.html()).toContain(`type="${type}"`);
    });

    it('should render the given theme', async () => {
      const theme = 'secondary';
      await wrapper.setProps({ theme });

      expect(wrapper.html()).toContain('bg-white');
    });

    it('should render the given size', async () => {
      const size = 'full';
      await wrapper.setProps({ size });

      expect(wrapper.html()).toContain('w-full');
    });

    it('should render the given disabled', async () => {
      const disabled = true;
      await wrapper.setProps({ disabled });

      expect(wrapper.html()).toContain('bg-neutral-100');
    });
  });
});