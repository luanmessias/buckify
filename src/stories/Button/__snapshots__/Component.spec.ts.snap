// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ButtonComponent > should render the default button 1`] = `
"<!--v-if-->
<button type=\\"button\\" class=\\"flex items-center justify-center gap-1 rounded-md text-sm h-14 transition-colors duration-200 ease-in-out bg-primary-500 hover:bg-primary-700 text-white w-40\\" theme=\\"primary\\" size=\\"medium\\"> Click me!</button>"
`;
